utility/service/service.go

```diff
@@ -70,17 +96,102 @@ func (s *servicer) HandleRelay(relay *coreTypes.Relay) (*coreTypes.RelayResponse
 		return nil, fmt.Errorf("Error admitting relay: %w", err)
 	}
 
-	// TODO: implement Persist Relay
-	// TODO: implement execution
-	// TODO: implement state maintenance
-	// TODO: validate the response from the node?
-	// TODO: (QUESTION) Should we persist SignedRPC?
-	return nil, nil
+	response, err := s.executeRelay(relay)
+	if err != nil {
+		return nil, fmt.Errorf("Error executing relay: %w", err)
+	}
+
+	// DISCUSS: should we validate the response from the node?
+	relayDigest, shouldStore, err := s.hasCollision(relay, response)
+	if err != nil {
+		return nil, fmt.Errorf("Error calculating relay service digest: %w", err)
+	}
+	if !shouldStore {
+		return response, nil
+	}
+
+	height := s.GetBus().GetConsensusModule().CurrentHeight()
+	writeCtx, err := s.GetBus().GetPersistenceModule().NewRWContext(int64(height))
+	if err != nil {
+		return nil, fmt.Errorf("Error getting a write context to update token usage for application %s: %w", relay.Meta.ApplicationAddress, err)
+	}
+	defer writeCtx.Release()
+
+	// DISCUSS: should we extend/use UnitOfWork for updating/retrieving token usage?
+	if err := writeCtx.RecordRelayService(relay.Meta.ApplicationAddress, relayDigest, relay, response); err != nil {
```

Comment:
IMPORTANT: @adshmh I unfortunately led you down the wrong path after our call yesterday.

- `readCtx` is used to read `on-chain` data (at a specific height)
	- We read on-chain data to understand the session quota (num tokens)

 -`writeCtx` is used to write to `on-chain` data (at a specific height)
	- We write on-chain data when the session is complete (e.g. need to send claim & proof transactions)

While the session is ongoing, we need to persist things (locally on disk or in memory but not on chain). Therefore, I think we might need to do something similar to `NonceDeduper` by leveraging `NewGenericFIFOSet` with a very large capacity.

I glossed over this but happy to go into more detail if need be. Just let me know.

Reply:
@Olshansk Thank you for the detailed review. Considering the requirements described in the relay mining paper, use of SMT backed by the `kvstore` package seems a better fit here. 
Please advise whether this seems correct.

Reply:
> use of SMT backed by the kvstore package seems a better fit here (while also steering clear of writeCtx

Correct.

>  In contrast, NewGenericFIFISet makes sense if the intention is to build/update the SMT only once,

Got it, makes sense.

Note that there will be a new SMT instance for every session. It's how v0 works and how v1 works. 

Unlike the trees for the state, where there are N trees always growing/changing, there is one new tree for every session. If a node is handling a couple of session at once, it will maintain a couple new SMTs until those session are claimed & proven.

Does that make sense?

Reply:
> Does that make sense?

Yes, thank you for the clarification.

> Unlike the trees for the state, where there are N trees always growing/changing, there is one new tree for every session. If a node is handling a couple of session at once, it will maintain a couple new SMTs until those session are claimed & proven.

Got it,  an SMT will be maintained for every session that has received service but has not been claimed yet.

Reply:
💯

Reply:
To add some context here the SMT writes are very efficient as we dont write to the database until `Commit()` is called so the writes are not very intensive until completed